src/widget_chooser.c

```diff
@@ -0,0 +1,502 @@
+					filter = gtk_file_filter_new();
+					gtk_file_filter_set_name(filter, patterns->line[count]);
+					gtk_file_filter_add_pattern(filter, patterns->line[count]);
+					gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(var->Widget), filter);
+				}
+				if (patterns) list_t_free(patterns);
+			}
+		}
+
+		/* Connect signals */
+
+		g_signal_connect(G_OBJECT(var->Widget), "file-activated",
+			G_CALLBACK(on_any_widget_file_activated_event), (gpointer)var->Attributes);
+		g_signal_connect(G_OBJECT(var->Widget), "selection-changed",
+			G_CALLBACK(on_chooser_widget_selection_changed_event), (gpointer)var->Attributes);
+		g_signal_connect(G_OBJECT(var->Widget), "update-preview",
+			G_CALLBACK(on_chooser_widget_update_preview_event), (gpointer)var->Attributes);
+		g_signal_connect(G_OBJECT(var->Widget), "current-folder-changed",
+			G_CALLBACK(on_any_widget_current_folder_changed_event), (gpointer)var->Attributes);
+		/* step: confirm-overwrite not implemented */
+	}
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Removeselected                                                      *
+ ***********************************************************************/
+
+void widget_chooser_removeselected(variable *var)
+{
+	gchar            *var1;
+	gint              var2;
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	fprintf(stderr, "%s(): Removeselected not implemented for this widget.\n",
+		__func__);
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Save                                                                *
+ ***********************************************************************/
+
+void widget_chooser_save(variable *var)
+{
+	FILE             *outfile;
+	GList            *element;
+	GtkTextBuffer    *buffer;
+	GtkTextIter       start, end;
+	gchar            *act;
+	gchar            *filename = NULL;
+	gchar            *text;
+	gchar            *string;
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	/* We'll use the output file filename if available */
+	act = attributeset_get_first(&element, var->Attributes, ATTR_OUTPUT);
+	while (act) {
+		if (strncasecmp(act, "file:", 5) == 0 && strlen(act) > 5) {
+			filename = act + 5;
+			break;
+		}
+		act = attributeset_get_next(&element, var->Attributes, ATTR_OUTPUT);
+	}
+
+	/* If we have a valid filename then open it and dump the widget's data to it */
+	if (filename) {
+		text = gtk_file_chooser_get_filename(GTK_FILE_CHOOSER(var->Widget));
+			if (text) {
+				string = g_strdup(text);
+				if ((outfile = fopen(filename, "w"))) {
+					fprintf(outfile, "%s", string);
+					fclose(outfile);
+				} else {
+					fprintf(stderr, "%s(): Couldn't open '%s' for writing.\n",
```

Comment:
Also log `strerror(errno)` maybe?

Reply:
Same code style as the rest of gtkdialog, which is scant on error messages.